zilliqa/src/consensus.rs

```diff
@@ -1056,11 +1109,61 @@ impl Consensus {
         // need those transactions again
         for tx in opaque_transactions {
             let account_nonce = self.state.get_account(tx.signer)?.nonce;
-            self.transaction_pool.insert_transaction(tx, account_nonce);
+            transaction_pool.insert_transaction(tx, account_nonce);
         }
         Ok(Some((proposal, broadcasted_transactions)))
     }
 
+    fn propose_new_block(&mut self) -> Result<Option<(Block, Vec<VerifiedTransaction>)>> {
+        let mut state = self.state.clone();
+        let mut tx_pool = self.transaction_pool.clone();
+        let mut votes = self.votes.clone();
+
+        let result = self.propose_new_block_at(&mut state, &mut tx_pool, &mut votes);
+
+        self.state = state;
+        self.transaction_pool = tx_pool;
+        self.votes = votes;
+
+        result
+    }
+
+    pub fn propose_pending_block(&self) -> Result<Option<Block>> {
+        let mut state = self.state.clone();
+        let mut tx_pool = self.transaction_pool.clone();
+        let mut votes = self.votes.clone();
+
+        let num = self.db.get_highest_block_number()?.unwrap();
+        let block = self.get_block_by_number(num)?.unwrap();
+
+        // If there are no votes for highest block then we have to create artificial supermajority including our vote only
+        // This is needed to satisfy aggregate signature (otherwise there's nothing agg signature can be built on)
+
+        if let Entry::Vacant(v) = votes.entry(block.hash()) {
+            let my_vote = Vote::new(
+                self.secret_key,
+                block.hash(),
+                self.public_key(),
+                self.view.get_view(),
+            );
+
+            let votes = (
+                vec![my_vote.signature()],
+                bitvec![u8, bitvec::order::Msb0; 1; 1],
+                0,
+                false,
+            );
+            v.insert(votes);
+        }
+
+        let result = self.propose_new_block_at(&mut state, &mut tx_pool, &mut votes)?;
```

Comment:
Haha, this is an `Option` so having it be named `result` made me do a double-take. Maybe naming it `proposal` will make it slightly more readable-at-first-glance?